src/creator.rs

```diff
@@ -97,53 +97,272 @@ impl<H: Hasher> Creator<H> {
+        n_members: NodeCount,
+    }
+
+    impl TestController {
+        fn new(
+            notifications_in: Receiver<NotificationOut<Hasher64>>,
+            max_units: usize,
+            n_members: NodeCount,
+        ) -> Self {
+            Self {
+                notifications_in,
+                units_out: vec![],
+                units: 0,
+                max_units,
+                candidates_by_round: vec![NodeMap::new_with_len(n_members)],
+                n_candidates_by_round: vec![NodeCount(0)],
+                n_members,
             }
         }
+
+        async fn control(&mut self) {
+            while self.units < self.max_units {
+                if let Some(NotificationOut::CreatedPreUnit(pre_unit)) =
+                    self.notifications_in.next().await
+                {
+                    self.units += 1;
+                    if self.n_candidates_by_round.len() <= pre_unit.round().into() {
+                        self.candidates_by_round
+                            .push(NodeMap::new_with_len(self.n_members));
+                        self.n_candidates_by_round.push(NodeCount(0));
+                    }
+                    if self.candidates_by_round[pre_unit.round() as usize][pre_unit.creator()]
+                        .is_none()
+                    {
+                        self.candidates_by_round[pre_unit.round() as usize][pre_unit.creator()] =
+                            Some([0; 8]);
+                        self.n_candidates_by_round[pre_unit.round() as usize] += NodeCount(1);
+                    }
+                    let full_unit = FullUnit::<Hasher64, Data>::new(
+                        pre_unit.clone(),
+                        Data::new(UnitCoord::new(0, 0.into()), 0),
+                        0,
+                    );
+                    for c in self.units_out.iter() {
+                        if c.unbounded_send(full_unit.unit()).is_err() {
+                            panic!("Failed to send a unit to a creator");
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn start(
+        n_members: usize,
+        n_fallen_members: usize,
+        max_units: usize,
+    ) -> (
+        TestController,
+        Vec<oneshot::Sender<()>>,
+        Vec<tokio::task::JoinHandle<()>>,
+        Sender<NotificationOut<Hasher64>>,
+    ) {
+        let session_id = 0;
+
+        let (to_test_controller, notifications_in) = mpsc::unbounded();
+
+        let mut test_controller = TestController::new(
+            notifications_in,
+            max_units,
+            (n_members + n_fallen_members).into(),
+        );
+
+        let mut handles = vec![];
+        let mut killers = vec![];
+
+        for node_ix in 0..n_members {
+            let (units_out, from_test_controller) = mpsc::unbounded();
+
+            let mut creator = Creator::new(
+                default_config(
```

Comment:
This is the default config for deployment, but not for tests (they quite slow because of this) -- take a look at testing::mock::gen_config.

Reply:
Done